compiler/rustc_codegen_ssa/src/mir/statement.rs

```diff
@@ -70,7 +76,15 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             mir::StatementKind::Intrinsic(box NonDivergingIntrinsic::Assume(ref op)) => {
                 if !matches!(bx.tcx().sess.opts.optimize, OptLevel::No | OptLevel::Less) {
                     let op_val = self.codegen_operand(bx, op);
-                    bx.assume(op_val.immediate());
+                    let imm = op_val.immediate();
+                    if let Some(value) = bx.const_to_opt_uint(imm) {
+                        if value == 0 {
+                            bx.unreachable();
+                            return ControlFlow::Break(());
+                        }
```

Comment:
```suggestion
                    if let Some(0) = bx.const_to_opt_uint(imm) {
                          bx.unreachable();
                          return ControlFlow::Break(());
```

Reply:
This looks expected. Submitting `assume(true)` doesn't make sense.

Reply:
Ah - you're right